recipes/openai_textcat.py

```diff
@@ -0,0 +1,470 @@
+    lang=("Language to use for tokenizer.", "option", "l", str),
+    model=("GPT-3 model to use for completion", "option", "m", str),
+    examples_path=("Examples file to help define the task", "option", "e", Path),
+    max_examples=("Max examples to include in prompt", "option", "n", int),
+    prompt_path=("Path to jinja2 prompt template", "option", "p", Path),
+    batch_size=("Batch size to send to OpenAI API", "option", "b", int),
+    segment=("Split sentences", "flag", "S", bool),
+    verbose=("Print extra information to terminal", "option", "flag", bool),
+)
+def textcat_openai_teach(
+    dataset: str,
+    input_path: Path,
+    labels: List[str],
+    chatgpt_bias: float = 0.5,
+    lang: str = "en",
+    model: str = "text-davinci-003",
+    batch_size: int = 10,
+    segment: bool = False,
+    examples_path: Optional[Path] = None,
+    prompt_path: Path = DEFAULT_PROMPT_PATH,
+    max_examples: int = 2,
+    verbose: bool = False,
+):
+    """Get bulk TextCat suggestions from an OpenAI API, using zero-shot or
+    few-shot learning.  The results can then be corrected using the
+    `textcat.manual` recipe.
+
+    Here, we use ChatGPT to suggest if a particular text talks about a recipe or
+    not. You can use the parameter `--chatgpt-bias` (float, 0 to 1 inclusive) to
+    set how much we prefer getting examples that ChatGPT thinks as recipes.
+
+    This approach lets you get the OpenAI queries out of the way upfront, which
+    can help if you want to use multiple annotators of if you want to make sure
+    you don't have to wait on the OpenAI queries.
+    """
+    api_key, api_org = _get_api_credentials(model)
+    examples = _read_prompt_examples(examples_path)
+    nlp = spacy.blank(lang)
+    if segment:
+        nlp.add_pipe("sentencizer")
+    openai = OpenAISuggester(
+        openai_model=model,
+        labels=labels,
+        max_examples=max_examples,
+        prompt_template=_load_template(prompt_path),
+        verbose=verbose,
+        segment=segment,
+        openai_api_key=api_key,
+        openai_api_org=api_org,
+    )
+    for eg in examples:
+        openai.add_example(eg)
+    stream = list(srsly.read_jsonl(input_path))
+    stream = openai(tqdm.tqdm(stream), batch_size=batch_size, nlp=nlp)
+    stream = _prefer_gpt(stream, chatgpt_bias)
+
+    return {
+        "dataset": dataset,
+        "view_id": "blocks",
+        "stream": stream,
+        "update": openai.update,
+        "config": {
+            "labels": openai.labels,
+            "batch_size": batch_size,
+            "exclude_by": "input",
+            "blocks": [
+                {"view_id": "classification"},
+                {"view_id": "html", "html_template": HTML_TEMPLATE},
+            ],
+            "show_flag": True,
+            "global_css": CSS_FILE_PATH.read_text(),
+        },
+    }
+
+
+def _prefer_gpt(stream, bias: float) -> Iterable[Dict]:
```

Comment:
My current implementation of the said sorter. The idea is for us to bias samples that ChatGPT thinks as positive.